Adds gear to create styled tables.

joaomcteixeira opened the pull request:
You are about to submit a new Pull Request. Before continuing make sure you read the [contributing guidelines](CONTRIBUTING.md) and you comply with the following criteria:

- [x] You have stick to Python. Talk with us before adding other programming languages to HADDOCK3
- [ ] Your PR is about CNS
- [x] Your code is well documented: proper docstrings and explanatory comments for those tricky parts
- [x] You structured the code into small functions as much as possible. You can use classes if there's a (state) purpose
- [x] code follows our coding style
- [x] You wrote tests for the new code
- [x] `tox` tests pass. *Run `tox` command inside the repository folder*
- [ ] `-test.cfg` examples execute without errors. *Inside `examples/` run `python run_tests.py -b`*
- [x] PR does not add any *install dependencies* unless permission granted by the HADDOCK team
- [x] PR does not break licensing
- [ ] Your PR is about writing documentation for already existing code :fire:
- [ ] Your PR is about writing tests for already existing code :godmode:

---

Add an independent `gear` python module to create human-readable and easily parsable tables. I implemented those now in `caprieval` but they can be used for any module.

Closes #442 

Tables now look like this:

``` 
             model                 caprieval_rank     score     irmsd      fnat    lrmsd     ilrmsd    dockq    cluster-id    cluster-ranking    model-cluster-ranking
../1_rigidbody/rigidbody_3.pdb           1           -32.675     1.791    0.361     3.575     2.642    0.541        -                -                     -          
../1_rigidbody/rigidbody_18.pdb          2           -26.322    11.042    0.111    17.990    18.247    0.104        -                -                     -          
../1_rigidbody/rigidbody_16.pdb          3           -25.486     3.050    0.361     5.909     4.081    0.410        -                -                     -          
../1_rigidbody/rigidbody_1.pdb           4           -24.790    10.926    0.111    18.156    18.059    0.103        -                -                     -          
../1_rigidbody/rigidbody_4.pdb           5           -24.462     3.802    0.194     6.879     5.712    0.311        -                -                     -          
../1_rigidbody/rigidbody_8.pdb           6           -24.325     9.164    0.056    16.579    12.117    0.097        -                -                     -          
../1_rigidbody/rigidbody_10.pdb          7           -23.595    10.410    0.028    17.466    15.835    0.080        -                -                     -          
../1_rigidbody/rigidbody_15.pdb          8           -23.535     1.385    0.500     2.601     1.999    0.651        -                -                     -          
../1_rigidbody/rigidbody_17.pdb          9           -20.917     4.543    0.139     7.955     6.410    0.257        -                -                     -          
../1_rigidbody/rigidbody_13.pdb          10          -19.378     8.200    0.056    14.108    12.063    0.118        -                -                     -          
../1_rigidbody/rigidbody_11.pdb          11          -18.528     3.541    0.222     6.474     4.438    0.336        -                -                     -          
../1_rigidbody/rigidbody_6.pdb           12          -15.703     5.595    0.056     9.893     7.205    0.182        -                -                     -          
../1_rigidbody/rigidbody_14.pdb          13          -15.399     8.973    0.083    15.060    13.387    0.117        -                -                     -          
../1_rigidbody/rigidbody_19.pdb          14          -15.378     5.019    0.028     9.609     4.415    0.183        -                -                     -          
../1_rigidbody/rigidbody_7.pdb           15          -14.898    10.100    0.028    16.580    14.345    0.086        -                -                     -          
../1_rigidbody/rigidbody_12.pdb          16          -14.364     9.725    0.056    15.688    14.463    0.102        -                -                     -          
../1_rigidbody/rigidbody_9.pdb           17          -13.154     2.657    0.417     5.361     3.950    0.458        -                -                     -          
../1_rigidbody/rigidbody_20.pdb          18          -11.732    10.836    0.000    17.827    17.082      nan        -                -                     -          
../1_rigidbody/rigidbody_2.pdb           19          -11.118    12.037    0.028    20.849    18.071    0.062        -                -                     -          
../1_rigidbody/rigidbody_5.pdb           20           -8.347     7.017    0.083    13.004    10.697    0.142        -                -                     -          
```

rvhonorato reviewed (changes requested):
Looks nice, but before it was also human readable. Also this will difficult integration with other analysis tools, when loading the "custom haddock-table" into a data-analysis platform which separator should be used? 

Since you are introducing yet another cutsom-format, this one should be thoroughly documented, also since the file has specific rules, should also have a specific name, how about `.hxt` for `haddock text file`?

rvhonorato commented:
Excel-like software can obviously identify the varying spaces; 

<img width="795" alt="Screenshot 2022-06-02 at 18 44 25" src="https://user-images.githubusercontent.com/9445814/171681688-a6f3bd01-cc98-48f7-9f52-c1426d9ab853.png">

but I did not have the same luck with pandas or R.

amjjbonvin commented:
Good point as we might want to have a module generating some plots at the end (as the 2.4 server does).

> but I did not have the same luck with pandas or R.
>

joaomcteixeira commented:
Thanks @rvhonorato for pointing out that this PR:

* creates an independent logic to create easy to read tables from dictionaries of data that can be used by any module inside and outside HADDOCK3. Thus, replacing the more simplistic, yet correct, tab-separated logic that was hardcoded in `caprieval`. Now `caprieval` uses this independent functionality.
* Does **not** change the data structure from `caprieval`. `caprieval` data structures created in #441 is **respected and maintained**.
* Adds thorough documentation on how to create tables in HADDOCK3 using this python module. Such documentation was missing in #441
* The "new" tables are space-separated values. Simply, variable width spacing is added to facilitate human readability.
* Adds back the logic that was incorporated in https://github.com/haddocking/haddock3/commit/6f7162767ea90aa95906c6027517309401f4251c but removed unnoticed in #230.

joaomcteixeira commented:
I added additional instructions on how to parse the tables. They are pure space-separated text tables. I believe, yet I might be wrong, that any person programming in any language would be able to parse them. Thanks for noting an example was missing :+1:

rvhonorato commented on src/haddock/gear/pretty_table.py:
```diff
@@ -0,0 +1,264 @@
+"""
+Tools to help the different modules creating pretty tables.
```
"pretty" is very subjective, how about "variable-width tables"?

rvhonorato commented on src/haddock/gear/pretty_table.py:
```diff
@@ -0,0 +1,264 @@
+      value3        value4        valut6
+
+
+These tables are space-separated values, where the number of spaces
+varies to make proper alignment of the text. If you are using Python,
+you can parse these tables with, for example:
```
```suggestion
These tables are space-separated values, where the number of spaces
varies to make proper alignment of the text, for purely aesthetic reasons. When parsing these tables one cannot rely on them following a regular spacing interval; the space between one column and the next might have 4 spaces, but then the next column may have 6, etc. Thus if these files are to be used for data analysis, the spacing must first be normalized. This can easily be done in Python, as seen in the example below. In other languages you might need to write a regular expression to achieve the same.
```

joaomcteixeira replied:
I didn't use this text exactly but I updated the documentation. I think it is clearer and clear now. Let me know otherwise.

rvhonorato commented on src/haddock/gear/pretty_table.py:
```diff
@@ -0,0 +1,264 @@
+    pass
+
+
+def create_human_readable_table(
```
same here, why not 'variable width'?

joaomcteixeira replied:
Updated the names.

rvhonorato commented on src/haddock/gear/pretty_table.py:
```diff
@@ -0,0 +1,264 @@
+        The formatted value to fill the table.
+    """
+    if isinstance(value, float):
+        return f"{value:.3f}".rjust(spacing, char)
```
since this function is supposed to be re-used, would make sense if one could this `.3f` was not hardcoded

joaomcteixeira replied:
Done

rvhonorato commented on src/haddock/modules/analysis/caprieval/capri.py:
```diff
@@ -387,11 +390,12 @@ def make_output(self):
         else:
             data["cluster-id"] = None
             data["cluster-ranking"] = None
-            data["self.model-cluster-ranking"] = None
```
oops

rvhonorato commented on src/haddock/modules/analysis/caprieval/capri.py:
```diff
@@ -645,7 +649,9 @@ def rearrange_ss_capri_output(
     else:
-        write_nested_dic_to_file(data, output_name)
+        data2 = convert_row_to_column_table(data)
+        table = create_human_readable_table(data2)
+        Path(output_name).write_text(table)
```
this should output its custom suffix also, how about:

```suggestion
        table, suffix = create_human_readable_table(data2)
        Path(output_name).with_suffix(suffix).write_text(table)
```

joaomcteixeira replied:
I disagree here because the tables python module I am implementing only concerns generating the formatted text. It's the responsibility of who/whatever asks for that text to save it with the extension found more appropriate. I have updated the extension in the `caprieval` tables to `txt`, though.

rvhonorato reviewed (commented):
A few other comments on top of my previous ones, see below;

amjjbonvin commented:
This is a discussion that should be taken offline …

joaomcteixeira commented:
Please have a look now:

- renamed the module to `tables` instead of `pretty_tables`
- added more documentation
- added functions to convert formats (not needed for now, but are there in case they are needed in the future)
- updated extensions of files output by `caprieval` module

joaomcteixeira commented:
I added a function to create tables with simple separators line space, comma, or tab from the data dictionary structure. I like the data dictionary structure as a pre-structure for tables. And we can use the tables gear to write any table format we may need.

mgiulini commented on tests/test_gear_tables.py:
```diff
@@ -0,0 +1,201 @@
+    assert result == table2
+
+
+def test_create_table_error():
+    d = {}
+    d["col1"] = list(range(5))
+    d["col2"] = ["haddock", "is", "amazing", None]
+    d["col3"] = [1.001, 123.543, 654.1, 98.999999]
+    with pytest.raises(TableFormatError):
+        create_adjusted_col_width_table(d)
+
+
+def test_create_table_error_2():
+    d = {}
+    d["col1"] = list(range(4))
+    d["col2"] = ["haddock", "is", "amazing"]
+    d["col3"] = [1.001, 123.543, 654.1, 98.999999]
+    with pytest.raises(TableFormatError):
+        create_adjusted_col_width_table(d)
```
These tests require a few words of description I would say

mgiulini commented on src/haddock/gear/tables.py:
```diff
@@ -0,0 +1,561 @@
+        rows[i] = column_spacing.join(rows[i])
+    table_data = os.linesep.join(rows)
+
+    # to situations to save tables with and header to avoid having extra
+    # empty lines.
```
This is not clear

mgiulini commented on src/haddock/gear/tables.py:
```diff
@@ -0,0 +1,561 @@
+
+    # to situations to save tables with and header to avoid having extra
+    # empty lines.
+    if header:
+        table = os.linesep.join([
+            header,
+            col_name_row,
+            table_data,
+            os.linesep,
+            ])
+    else:
+        table = os.linesep.join([
+            col_name_row,
+            table_data,
+            os.linesep,
+            ])
```
I think it's better to create the overall table first:
```
table = os.linesep.join([
            col_name_row,
            table_data,
            os.linesep,
            ])
```
and then check for the existence of the header: shorter code, better readability, do you agree?

mgiulini reviewed (commented):
@joaomcteixeira personally I don't like this implementation. I mean, the code looks fine, but I see a major critical point.

This PR adds about a thousand lines of code for a purpose (creating human-readable and easily parsable tables) that is not worth the high costs this code introduces in terms of long-time maintenance and readability. Issue #458 is an example in this context: previously it would have taken two minutes to correct the format of each value, now a developer new to the code would need to dive into (and probably modify) the gear function for a lot of time to figure out how to close this (fairly trivial) issue.

In my opinion it would be better if each module had its own function to output human readable tables, without overloading the gear. It would be easier to correct things, to implement new features and to lower maintenance costs.

rvhonorato commented:
Since this is taking a direction of becoming "the way of outputting tables with haddock3" and not just an aesthetic whim, I have to agree with @mgiulini that this implementation, (also) from my point of view, is not worth in terms of long-time maintenance and readability. 

I'd even suggest that before we code our own version of gears that have no scientific impact, we first look into well consolidated packages with high usage that can do trivial tasks such as these and that will both not increase the learning curve or take up our development time. 

I am against this implementation but I'm open to taking a vote if it comes to that.
